### PR TITLE
Default and python configs updated at compilation

### PR DESCRIPTION
**Pull request by bobbyshermi**
Fix issue: https://github.com/Avaiga/taipy-config/issues/69

- Make the _applied_config and the _default_config two different instances.
- Copy sections instead of re-using the same instance in the _Config._update method
- make the _config initializations consistent
- Create clear methods to clear the _applied_config before compilation.
- Unit tests

Note the Section Class changed. A new abstract method must be implemented. The fix must be propagated to taipy-core

### CODE REVIEW

**Review by trgiangdo — CHANGES_REQUESTED** [review]
Thank you for your Pull Request. This is a severe bug and we are so happy that you found out and propose to fix it.

For the code change, I have some concerns. Please address these.
1. I don't see the purpose of the `clear()` method. The `_applied_config` and the `_default_config` are already 2 different instances (since `_Config._default_config()` returns a different `_Config` instance each call). Your change on copy section and unique_section when updating seems enough. Please clarify this for us.
2. The PR contains several commits of your previous PR, which will be merge soon. Before merging this PR, please rebase the branch with develop. And next time, can you use a fresh branch from develop? It will make it much easier for us to review. Thank you.

File: src/taipy/config/_config.py [hunk 1]
```diff
@@ -25,6 +25,14 @@ def __init__(self):
         self._unique_sections: Dict[str, UniqueSection] = {}
         self._global_config: GlobalAppConfig = GlobalAppConfig()
 
+    def clear(self):
+        self._global_config.clear()
+        for unique_section in self._unique_sections.values():
+            unique_section.clear()
+        for sections in self._sections.values():
+            for section in sections.values():
+                section.clear()
```
Comment (jrobinAV) [inline src/taipy/config/_config.py]: I prefer the name `clean` as it does not make the _config empty, but it cleans all the elements.
Plus, can you make it hidden?

```suggestion
    def _clean(self):
        self._global_config._clean()
        for unique_section in self._unique_sections.values():
            unique_section._clean()
        for sections in self._sections.values():
            for section in sections.values():
                section._clean()
```

File: src/taipy/config/config.py [hunk 2]
```diff
@@ -226,7 +226,7 @@ def _override_env_file(cls):
     @classmethod
     def __compile_configs(cls):
         Config._override_env_file()
-        cls._applied_config = _Config._default_config()
+        cls._applied_config.clear()
```
Comment (jrobinAV) [inline src/taipy/config/config.py]:
```suggestion
        cls._applied_config._clean()
```

File: src/taipy/config/global_app/global_app_config.py [hunk 3]
```diff
@@ -135,6 +135,13 @@ def default_config(cls) -> GlobalAppConfig:
         config._repository_properties = cls._DEFAULT_REPOSITORY_PROPERTIES.copy()
         return config
 
+    def clear(self):
```
Comment (jrobinAV) [inline src/taipy/config/global_app/global_app_config.py]:
```suggestion
    def _clean(self):
```

File: src/taipy/config/section.py [hunk 4]
```diff
@@ -39,6 +39,10 @@ def __copy__(self):
     def name(self):
         raise NotImplementedError
 
+    @abstractmethod
+    def clear(self):
```
Comment (jrobinAV) [inline src/taipy/config/section.py]:
```suggestion
    def _clean(self):
```

File: tests/config/utils/section_for_tests.py [hunk 5]
```diff
@@ -38,6 +38,10 @@ def attribute(self):
     def attribute(self, val):
         self._attribute = val
 
+    def clear(self):
```
Comment (jrobinAV) [inline tests/config/utils/section_for_tests.py]:
```suggestion
    def _clean(self):
```

File: tests/config/utils/unique_section_for_tests.py [hunk 6]
```diff
@@ -39,6 +39,10 @@ def attribute(self):
     def attribute(self, val):
         self._attribute = val
 
+    def clear(self):
```
Comment (jrobinAV) [inline tests/config/utils/unique_section_for_tests.py]:
```suggestion
    def _clean(self):
```

**Review by jrobinAV — CHANGES_REQUESTED** [review]

**Comment by bobbyshermi** [comment]
> Thank you for your Pull Request. This is a severe bug and we are so happy that you found out and propose to fix it.
> 
> For the code change, I have some concerns. Please address these.
> 
> 1. I don't see the purpose of the `clear()` method. The `_applied_config` and the `_default_config` are already 2 different instances (since `_Config._default_config()` returns a different `_Config` instance each call). Your change on copy section and unique_section when updating seems enough. Please clarify this for us.
> 2. The PR contains several commits of your previous PR, which will be merge soon. Before merging this PR, please rebase the branch with develop. And next time, can you use a fresh branch from develop? It will make it much easier for us to review. Thank you.

Hello, 

1- `_clean` is used to keep all sections inside `_applied_config` the same instances between two compilations. Otherwise, some tests are failing because at the registration time you do expose `_applied_config._sections['name']` as an attribute. So If the instance changes, the attribute is not dynamically updated and still points to the old instance which can be out of date.

2- I believe the merge of my previous PR solved the issue. I will branch from Avaiga:develop branch next time. Thx.

Thank you for the review.

**Review by jrobinAV — APPROVED** [review]
Thx for the update.

**Comment by trgiangdo** [comment]
> `_clean` is used to keep all sections inside `_applied_config` the same instances between two compilations

That's great. I would love to have a more generic way to implement the `_clear` method for all sections. But for now, this will do.

Can you also do a PR on taipy-core to propagate the `_clear()` method to other sections and unique sections?

**Comment by bobbyshermi** [comment]
> > `_clean` is used to keep all sections inside `_applied_config` the same instances between two compilations
> 
> That's great. I would love to have a more generic way to implement the `_clear` method for all sections. But for now, this will do.
> 
> Can you also do a PR on taipy-core to propagate the `_clear()` method to other sections and unique sections?

I can propagate the changes if you want, but I won't have time this week. I mostly have time to work on taipy on the week-ends.

Thx,

**Comment by trgiangdo** [comment]
Then I propose we delay the merge of this PR, since it will cause failing tests in other Taipy repositories.

Thanks for your contribution 🫶

**Comment by trgiangdo** [comment]
I just re-reviewed this PR and I have 1 more remark.

Essentially, the `_Config._clean()` method will reset all attributes of all sections and unique_sections to some sort of default value (like the SectionForTest).
And when compiling the `_applied_config`, these sections and unique_sections are replaced by instances in _default_config and _python_config anyway.

So, maybe instead of clean each one, we just remove these sections and unique_sections from the dictionary when calling `_clean()`?
Then we will be able to avoid the need for propagate the `_clean()` method to all sections, which will be hard to maintain.

@jrobinAV @bobbyshermi let me know your thoughts about this. Thank you

**Comment by jrobinAV** [comment]
@trgiangdo I believe it does not work to remove sections since before and after the compilation the instances must be the same.

**Comment by trgiangdo** [comment]
When we compile the config, we remove those anyway (by replacing it with copies of sections and unique_sections from _default_config), no?

**Comment by jrobinAV** [comment]
Not really... We update each attribute one by one. The purpose is to have the following behavior:
```
config_1 = Config.configure_data_node("foo", "bar")
config_2 = Config.configure_data_node("foo", "baz")
assert config_1 is config_2
```
In the first line, config_1 references `Config._applied_config._sections["foo"]`.
Then before configuring `config_2` we clean the `_applied_config`.
Then in the second line, we make config_2 also a reference of `Config._applied_config._sections["foo"]`.
Since we want `config_1` and `config_2` to be the same instance, we must update the attributes, but not the instance itself.

I don't know if I am clear.

**Review by trgiangdo — APPROVED** [review]
The `_clean()` method has been propagated to all other Section and UniqueSection children.
This PR is ready to be merged.

**Review by trgiangdo — APPROVED** [review]
The `_clean()` method has been propagated to all other Section and UniqueSection children.
This PR is ready to be merged.

**Pull request merged**